jax/interpreters/pxla.py

```diff
@@ -298,29 +280,22 @@ class ShardedDeviceArray(xla.DeviceArray):
   that is, it is only an optimization to reduce transfers.
 
   The number of device buffers underlying a ShardedDeviceArray instance is equal
-  to the number of replicas of the computation that produced it. Each buffer
-  represents a shard of the original array, meaning a slice along its leading
-  axis. These component buffers reside on distinct devices, but need not
-  represent distinct logical shards. The correspondence can be computed with
-  the assign_shards_to_replicas function.
+  to the number of devices of the computation that produced it. Each buffer
+  represents a shard of the original array, indicated by its corresponding index
+  into the logical array. These component buffers reside on distinct devices,
+  but need not represent distinct logical shards.
   """
-  __slots__ = ["device_buffers", "axis_size"]
-  _collect = staticmethod(onp.stack)
+  __slots__ = ["device_buffers", "logical_indices"]
 
-  def __init__(self, aval, device_buffers):
+  def __init__(self, aval, logical_indices, device_buffers):
+    assert len(logical_indices) == len(device_buffers)
```

Comment:
This is a good idea. I'm gonna hold off for now since it'll be easy to add in a follow-up PR, and we might have a better idea of what makes sense later (plus I should probably write a benchmark to measure SDA creation time...).

Reply:
Why is it important to support the rich indexing provided by NumPy? Are these logical_indices arising from NumPy indexing or from some other simpler API?